overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -336,18 +336,337 @@ func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []st
 	return filePerms, nil
 }
 
-var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+var (
+	// The following must be escaped as literals in a path pattern:
+	problematicChars = `\[\]`
+	// A single safe or escaped unsafe char in a path
+	safePathChar = fmt.Sprintf(`([^%s]|\\[%s])`, problematicChars, problematicChars)
+
+	// The following matches valid path patterns
+	allowablePathPattern       = fmt.Sprintf(`^/%s*$`, safePathChar)
+	allowablePathPatternRegexp = regexp.MustCompile(allowablePathPattern)
+)
+
+// Expands all groups in the given path pattern. Groups are enclosed by '{' '}'.
+// Returns a list of all the expanded path patterns, or an error if the given
+// pattern is invalid.
+func ExpandPathPattern(pattern string) ([]string, error) {
+	patternLen := len(pattern)
+	expanded := defaultPrefixes
+	currLiteralStart := 0
+	for i := 0; i < patternLen; i++ {
+		if pattern[i] == '\\' {
+			i += 1
+			if i == patternLen {
+				return nil, fmt.Errorf(`invalid path pattern: trailing unescaped '\' character: %q`, pattern)
+			}
+			continue
+		}
+		if pattern[i] == '}' {
+			return nil, fmt.Errorf(`invalid path pattern: unmatched '}' character: %q`, pattern)
+		}
+		if pattern[i] != '{' {
+			continue
+		}
+		infix := pattern[currLiteralStart:i]
+		groupExpanded, groupEnd, err := expandPathPatternFromIndex(pattern, i+1, patternLen)
+		if err != nil {
+			return nil, err
+		}
+		newExpanded := make([]string, 0, len(expanded)*len(groupExpanded))
+		for _, prefix := range expanded {
+			for _, suffix := range groupExpanded {
+				newExpanded = append(newExpanded, prefix+infix+suffix)
+			}
+		}
+		expanded = newExpanded
+		currLiteralStart = groupEnd + 1
+		i = groupEnd // let for loop increment to index after '}'
+	}
+	if len(expanded) == 1 && expanded[0] == "" {
+		// Didn't expand any groups, so return whole pattern.
+		return []string{cleanPattern(pattern)}, nil
+	}
+	// Append trailing literal string, if any, to all previously-expanded
+	// patterns, and clean the resulting patterns.
+	alreadySeen := make(map[string]bool, len(expanded))
+	newExpanded := make([]string, 0, len(expanded))
```

Comment:
I think it may be possible to modify `expanded` in place here

Reply:
It is possible, but then there would be lots of re-allocations if repeated appending exceeds the capacity multiple times. I'm essentially trying to grow the capacity of `expanded`, and afaik, the best way to do so is to allocate a new slice, `copy()`, then reassign the original variable to the new slice.

I'll move `expanded = newExpanded` to right after the `copy()`, and then append to `expanded`.

Reply:
Ah there's a secret pattern here, if you use 
```go
expanded = append(expanded, make([]string, len(currPrefixes))
```
Go will recognise this and convert it into a single call to `runtime.GrowSlice(len(currPrefixes))`